test/coverage/utils/ExpectedTest.cpp

```diff
@@ -23,16 +29,361 @@ class TestFixture : public testing::Test {
 
 	// Run once per execution of the test application.
 	// Used for setup of all tests. Has access to this instance.
-	TestFixture() = default;
-	~TestFixture() = default;
+	ExpectedTest() = default;
+	~ExpectedTest() = default;
 
 	// Run once per execution of the test application.
 	// Used only for global setup outside of tests.
 	static void SetUpTestSuite() {}
 	static void TearDownTestSuite() {}
 };
 
-// TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+uint64_t get_rand() {
+	static std::random_device rd;
+	static std::mt19937 mt(rd());
+	static std::uniform_int_distribution<int> dist(0, 1 << 30);
+	return dist(mt);
+}
+
+TEST_F(ExpectedTest, ExpectScalarScalar) {
+	using namespace std;
+
+	auto sample = get_rand();
+	auto lfn = [&]() -> Expected<int, int> { return sample; };
+	auto expected = lfn();
```

Comment:
Is this doing something specific? Wouldn't regular Expected construction be adequate? What does this test that isn't covered with:
```suggestion
	Expected<int, int> expected(sample);
```

Reply:
This test proves that the implementation can handle expected and unexpected types that are identical. The motivation for having to tag the unexpected type is driven by the need to deal with this possibility.

I think the std::expected thing is interesting, but would be better if the unexpected type where maintained to be something derived from an exception, if the calling code wanted to easily default back to exception behaviors/patterns. Having done all this work, I don't see this as being worth the trouble. It is a dressed up variant, and more code that could break.

Reply:
Could that not be proved by passing parameters to the regular constructors of `Expected`? It looks to me that the lambda is unnecessary indirection here.

What is the benefit of
```
int sample = get_rand();
auto lfn = [&]() -> Expected<int, int> { return Unexpected(sample); };
auto expected = lfn();
```
over
```
int sample = get_rand();
Expected<int, int> expected(Unexpected(sample));
```

Reply:
It seems to me mocking the function return behavior is closer to the real world behavior, and the syntax overhead is minimal. It may be the case that the lambda function is effectively inlined by the compiler. Who knows? I'm not sure that it matters. These unit tests are miles away from actually meaningful tests where integration is actually tested. I don't care either way. If you want them to not use lambda functions, I'll rip all the lambda functions out.

Is that the ask?

Reply:
I'm assuming that you want the lambda functions ripped out. I will start to work on that. However, in tests like this, there is no meaningful 'optimization'. Unit tests don't prove much, and 'optimizing' seems like raking leaves.

Reply:
In the three years I spent writing the testing framework for the ADIM3 interface to the A100, I never wrote a gtest. I used ROS/python Lunar Loggerhead interface, pybind11 for C++ encapsulation, pcapplusplus for packet assembly and spoofing, threads and condition variables for packet timing, OpenSSL for mTLS exchange, rpyc for dspace integration, and I was hoping to work on NeoVi Fire2 Linux libraries for live CAN capture.

My test specs were written in YAML, and were imported to python with ruamel.YAML. Pybind11 gave my C++ objects for packet description, timer chain objects, etc. python presence. I used nlohman::json for CBOR serialization and deserialization of variable length records, and fixed sized record index file for the ability to reel-rock the ethernet+ROS topic combined file stream.

I had written python code to parse the ARXML CAN message database, and use the enum values in the YAML driven testing specs.

I never considered gtests to be useful.

All that stuff I wrote is the main line stuff that Cruise is still using to test firmware drops. GM is also using it to test both A100 and A110 for ethernet behavior.

Reply:
The thing I hated most about the Cruise situation was that Lunar Loggerhead was six years old, and didn't have a working message reflection interface. This is something the ROS folks fixed. I was hoping that such a thing would be fixed for this project, but that does not seem to be the case. I have code that proves it for all of the current uProtocol fields for everything but the details field. Fortunately, there seem to be no other repeated or any use cases, but even though it works and fills an obvious need, its dead code, never to be used.

Reply:
> I'm assuming that you want the lambda functions ripped out. I will start to work on that. However, in tests like this, there is no meaningful 'optimization'. Unit tests don't prove much, and 'optimizing' seems like raking leaves.

@debruce - I want to be able to read the test and know what is being tested vs what is just extra stuff. It seems to me the lambdas are not adding anything here in terms of the testing. Instead they only add cognitive load for the reader. They might be useful if they were doing some sequence of tasks that needed to be reused throughout a test, but all they are doing is adding extra steps to constructing an object - that construction is the real purpose of the test.